doc/manual/src/release-notes/rl-2.4.md

```diff
@@ -1,8 +1,492 @@
+
+    * `nix profile history` shows what packages were added, upgraded
+      or removed between each version of a profile.
+
+    * `nix profile diff-closures` shows the changes between the
+      closures of each version of a profile. This allows you to
+      discover the addition or removal of dependencies or size
+      changes.
+
+    **Warning**: after a profile has been updated using `nix profile`,
+    it is no longer useable with `nix-env`.
+
+  - `nix store diff-closures` shows the differences between the
+    closures of two store paths in terms of the versions and sizes of
+    dependencies in the closures.
+
+  - `nix store make-content-addressable` rewrites an arbitrary closure
+    to make it content-addressed. Such paths can be copied into other
+    stores without requiring signatures.
+
+  - `nix bundle` uses the [`nix-bundle`
+    program](https://github.com/matthewbauer/nix-bundle) to convert a
+    closure into a self-extracting executable.
+
+  - Various other replacements for the old CLI, e.g. `nix store gc`,
+    `nix store delete`, `nix store repair`, `nix nar dump-path`, `nix
+    store prefetch-file`, `nix store prefetch-tarball`, `nix key` and
+    `nix daemon`.
+
+* Nix now has an **evaluation cache** for flake outputs. For example,
+  a second invocation of the command `nix run nixpkgs#firefox` will
+  not need to evaluate the `firefox` attribute because it's already in
+  the evaluation cache. This is made possible by the hermetic
+  evaluation model of flakes.
+
+* The new `--offline` flag disables substituters and causes all
+  locally cached tarballs and repositories to be considered
+  up-to-date.
+
+* The new `--refresh` flag causes all locally cached tarballs and
+  repositories to be considered out-of-date.
+
+* Many `nix` subcommands now have a `--json` option to produce
+  machine-readable output.
+
+* `nix repl` has a new `:doc` command to show documentation about
+  builtin functions (e.g. `:doc builtins.map`).
+
+* Binary cache stores now have an option `index-debug-info` to create
+  an index of DWARF debuginfo files for use by
+  [`dwarffs`](https://github.com/edolstra/dwarffs).
+
+* To support flakes, Nix now has an extensible mechanism for fetching
+  source trees. Currently it has the following backends:
+
+  * Git repositories
+
+  * Mercurial repositories
+
+  * GitHub and GitLab repositories (an optimisation for faster
+    fetching than Git)
+
+  * Tarballs
+
+  * Arbitrary directories
+
+  The fetcher infrastructure is exposed via flake input specifications
+  and via the `fetchTree` built-in.
+
+* **Languages changes**: the only new language feature is that you can
+  now have antiquotations in paths, e.g. `./${foo}` instead of `./. +
+  foo`.
+
+* **New built-in functions**:
+
+  - `builtins.fetchTree` allows fetching a source tree using any
+    backends supported by the fetcher infrastructure. It replaces
+    `fetchGit`, `fetchMercurial` and `fetchTarball`.
```

Comment:
I guess it's worth mentioning that `fetchGit`/etc are still available, but considered deprecated (I guess?)

Reply:
Removing builtins is a bad idea because old expressions won't be buildable anymore. Being able to build expressions from ages ago is a remarkable and unique capability.
In Nixpkgs we don't have this constraint, because legacy software can pin an older version but still be integrated into a system based on `master`, but when the Nix language or its semantics change, this can't be done.
Deprecation would imply removal. Otherwise, we're just bugging people for no real reason.

Reply:
For the same reasons, I'm very concerned about the unreproducible behavior of `builtins.fetchGit` and/or `fetchTree`. Too little consideration is given to the smudge filters, running them only in arbitrary circumstances which [change](https://github.com/NixOS/nix/issues/5260), while the only sane default is not to run them unless specific supported ones are explicitly requested, such as git-lfs.

Reply:
Agreed. I just wanted to point out that `It replaces fetchGit, [...]` sounds exactly like it's not available anymore (to me at least, though I'm not a native speaker).

Btw, I fully agree that builtins shouldn't be removed because old Nix expressions should remain evaluatable.

Reply:
@roberth Make an issue about it, the release notes are not the place to discuss this.

Reply:
I agree, so I've created #5313. I commented here because it would catch your attention, as I did raise the issue before, albeit in a PR.

Reply:
> Deprecation would imply removal. Otherwise, we're just bugging people for no real reason.

I was wrong about this. `fetchGit` should remain untouched, to preserve some degree of reproducibility, but it is broken so it _should_ be avoided, except when building legacy software, in which case a warning isn't too much of a nuisance.

Reply:
Changed the wording, there is no intent to remove `fetch*`.